client/http/http.go

```diff
@@ -89,7 +89,6 @@ func (tc *TracedClient) Do(ctx context.Context, req *http.Request) (*http.Respon
 	if err != nil {
 		ext.Error.Set(ht.Span(), true)
```

Comment:
In order to make the rest of the code easier to understand (remove else and early exits) i would suggest the following:
- We should return from the function after the error in line 91, so the rest of the code is the else block.
- If the Do function does not return an error then the response is most probably always non nil so there is no need to check.

Reply:
This was our original approach, but we've encountered issues with the tracing.

Apparently, the `do` function must be executed at least once before we can set any spans. This means we cannot  `ext.HTTPMethod.Set(ht.Span(), ...)` before a `do`, which means returning early will not record the spans properly, unless we duplicate the `ext` calls both in the `if err != nil` and in the following flow.

Not sure if that makes sense :) Maybe @tpaschalis can explain it better?

Reply:
@mantzas 

This is what it could look like, with the early return (comments for clarity).

```
rsp, err := tc.do(req)
if err != nil {
  ext.HTTPMethod.Set(ht.Span(), req.Method) // don't forget to set these, after `do`
  ext.HTTPUrl.Set(ht.Span(), req.URL.String())
  ext.Error.Set(ht.Span(), true)

  return rsp, err
}

ext.HTTPMethod.Set(ht.Span(), req.Method) // don't forget to set these, after `do`
ext.HTTPUrl.Set(ht.Span(), req.URL.String())
ext.HTTPStatusCode.Set(ht.Span(), uint16(rsp.StatusCode))

reqDurationMetrics.
  WithLabelValues(req.Method, req.URL.Host, strconv.Itoa(rsp.StatusCode)).
  Observe(time.Since(start).Seconds())

if hdr := req.Header.Get(encoding.AcceptEncodingHeader); hdr != "" {
  rsp.Body = decompress(hdr, rsp)
}

return rsp, err
```

Reply:
option no2 seems better. 
you can even set the
```ext.HTTPMethod.Set(ht.Span(), req.Method) // don't forget to set these, after `do`
ext.HTTPUrl.Set(ht.Span(), req.URL.String())```
before you actually call the do method and you need probably to call also
``` ext.Error.Set(ht.Span(), false)``` on success.

Reply:
You _cannot_ call any of the `ext` methods before the `do` - the tracer is not properly configured until after the `RoundTripper` first gets invoked. So the two `ext` calls will have to remain _after_ the `do`.

Let me know if we're still happy with that, I'll update accordingly.

Reply:
this is odd since they seems not to be related with the call but yes, no problem.